libs/ml/include/ml/graph.hpp

```diff
@@ -56,7 +56,8 @@ class Graph
   virtual ~Graph() = default;
 
   ArrayType    Evaluate(std::string const &node_name, bool is_training = true);
-  void         BackPropagate(std::string const &node_name, ArrayType const &error_signal);
+  void         BackPropagateSignal(std::string const &node_name, ArrayType const &error_signal);
```

Comment:
these names are confusing. If i understand correctly wouldn't it be better to rename as follows:
BackPropagate -> BackPropagateLoss
BackPropagate -> BackPropagateSignal